src/device/usbd.c

```diff
@@ -543,9 +548,18 @@ static bool process_control_request(uint8_t rhport, tusb_control_request_t const
 
           dcd_set_config(rhport, cfg_num);
 
-          if ( !_usbd_dev.configured && cfg_num ) TU_ASSERT( process_set_config(rhport, cfg_num) );
+          if (0 == cfg_num)
+          {
+            usbd_reset_config(rhport);
+            if (tud_umount_cb) tud_umount_cb();
+          }
+          else
+          {
+            TU_VERIFY( (0 == _usbd_dev.selected_config) || (cfg_num == _usbd_dev.selected_config) );
```

Comment:
I am not sure why we need `cfg_num == _usbd_dev.selected_config` in the TU_VERIFY(), if it is the case, we will just skip doing anything since host SET the same config. If they are different and both are not zero, we will purposely invoke `dcd_set_config(rhport, 0)` first for dcd to close all opened endpoints (as mentioned above), then call `dcd_set_config(rhport, cfg_num)` again, `then process_set_config()`

Again: this requires all DCD to implement dcd_set_config(0) as close all endpoints.

Reply:
It follows the USB spec.  "If the specified configuration value matches the configuration value from a configuration descriptor, then that configuration is selected and the device remains in the Configured state. Otherwise, the device responds with a Request Error."  A "Request Error" is a stall, and we ostensibly get this by TU_VERIFY returning a false (same as the default 'Unknown/Unsupported request' case).

Reply:
I think the specs mean if the config num == the one in the descriptors, then use that config descriptor.  `cfg_num == _usbd_dev.selected_config` means host SET the same config value as I did before.